arbiter/drivers/onedrive.cpp

```diff
@@ -0,0 +1,292 @@
+        if (obj.contains("@odata.nextLink"))
+            pageUrl = obj.at("@odata.nextLink");
+        else
+            break;
+
+    } while (true);
+
+    return result;
+}
+
+std::vector<std::string> OneDrive::glob(std::string path, bool verbose) const
+{
+    path.pop_back();
+    bool recursive(path.back() == '*');
+    if (recursive)
+        path.pop_back();
+
+    if (path.back() == '/')
+        path.pop_back();
+
+    return processList(path, recursive);
+}
+
+bool OneDrive::get(
+        const std::string path,
+        std::vector<char>& data,
+        const http::Headers userHeaders,
+        const http::Query query) const
+{
+    Resource resource(path);
+    http::Headers headers(m_auth->headers());
+    headers["Content-Type"] = "application/octet-stream";
+    headers.insert(userHeaders.begin(), userHeaders.end());
+
+    drivers::Https https(m_pool);
+    const auto res(https.internalGet(resource.binaryEndpoint(), headers));
+
+    if (!res.ok()) {
+        std::cout <<
+            "Failed get - " << res.code() << ": " << res.str() << std::endl;
+        return false;
+    }
+
+    data = res.data();
+    return true;
+}
+
+OneDrive::Auth::Auth(std::string s) {
+    const auto config = json::parse(s);
+    m_token = config.at("access_token").get<std::string>();
+    m_refresh = config.at("refresh_token").get<std::string>();
+    m_redirect = config.at("redirect_uri").get<std::string>();
+    m_id = config.at("client_id").get<std::string>();
+    m_secret = config.at("client_secret").get<std::string>();
+    m_expiration = Time().asUnix();
+}
+
+std::unique_ptr<OneDrive::Auth> OneDrive::Auth::create(const std::string s)
+{
+    return makeUnique<Auth>(s);
+}
+
+void OneDrive::Auth::refresh()
+{
+    //only refresh if we get within 2 minutes of the token ending
+    const auto now(Time().asUnix());
+    if (m_expiration - now > 120)
+        return;
+
+    http::Pool pool;
+    drivers::Https https(pool);
+    http::Headers headers({ });
+
+    headers["Accept"] = "application/json";
+    headers["Content-Type"] = "application/x-www-form-urlencoded";
+
+    json bodyJson = {
+        { "access_token", m_token },
+        { "refresh_token", m_refresh },
+        { "client_id", m_id },
+        { "client_secret", m_secret },
+        { "scope", "offline_access+files.read.all+user.read.all" },
+        { "grant_type", "refresh_token" }
+    };
+
+    //create url-encoded body
+    std::string body = "";
+    for (auto& i: bodyJson.items())
+        body += i.key() + "=" + i.value().get<std::string>() + "&";
+    body.pop_back();
+    std::vector<char> content(body.begin(), body.end());
+
+    const auto res(https.internalPost(this->getRefreshUrl(), content, headers));
+    const auto response(json::parse(res.str()));
+
+    //reset the token, refresh, and expiration time
+    m_token = response.at("access_token").get<std::string>();
+    m_refresh = response.at("refresh_token").get<std::string>();
+    m_expiration = now + 3599;
```

Comment:
Looks like we're guessing at the expiration, but there should be an expiration date in the response.

Reply:
There is, but it's always 3599

Reply:
What does this mean exactly?  Isn't there an actual timestamp as the expiration in the response?  Maybe paste a sample response here.

Reply:
It's not a timestamp, it's just a number in seconds that represents how far in the future this token will expire. Maybe should still pull from the json for 3599 itself just to be consistent.
```
{
    "access_token": "eyJ0eXAiOiJKV1QiLCJhbGciOiJSUzI1NiIsIng1dCI6Ik5HVEZ2ZEstZnl0aEV1Q...",
    "token_type": "Bearer",
    "expires_in": 3599,
    "scope": "user.read%20mail.read",
    "refresh_token": "AwABAAAAvPM1KaPlrEqdFSBzjqfTGAMxZGUTdM0t4B4...",
}
```
From https://docs.microsoft.com/en-us/graph/auth-v2-user#5-use-the-refresh-token-to-get-a-new-access-token

Reply:
Sounds good, then just replace the magic number 3599 with the `expires_in` value from the response as you mentioned.